File: Hash_Table/valid_anagram.py
""" 
Given two strings s and t, return true if t is an anagram of s, and false otherwise.

An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all the original letters exactly once.

Example :
Input: s = "anagram", t = "nagaram"
Output: true

Input: s = "rat", t = "car"
Output: false
"""


class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False

        apl_dic: dict[str, int] = {}
        for s_ in s:
            exist_val = apl_dic.get(s_)
            if exist_val:
                apl_dic[s_] = exist_val + 1
                continue

            apl_dic[s_] = 1

        for t_ in t:
            exist_val = apl_dic.get(t_)
            if not exist_val:
                return False

            apl_dic[t_] = exist_val - 1

        return True


if __name__ == "__main__":
    s1, t1 = "anagram", "nagaram"
    solution = Solution()
    result = solution.isAnagram(s1, t1)
    print(result)

File: Hash_Table/valid_anagram.go
package main

import "fmt"

func isAnagram(s, t string) bool {
	s_rune := []rune(s)
	t_rune := []rune(t)
	if len(s_rune) != len(t_rune) {
		return false
	}

	contains := make(map[rune]int)
	for _, v := range s_rune {
		// fmt.Printf("%v - %T\n", v, v)
		count, exist := contains[v]
		if exist {
			contains[v] = count + 1
		} else {
			contains[v] = 1
		}
	}
	fmt.Println(contains)

	for _, v := range t_rune {
		count, exist := contains[v]
		if !exist || count <= 0 { 
			return false
		}
		contains[v] = count - 1
	}

	return true

}

func main() {
	s := "anagram"
	t := "nagaram"
	result := isAnagram(s, t)
	fmt.Println(result)

}
